### DISCUSSION TITLE
Hardware UART: question about setFIFOSize()

### DISCUSSION THREAD

**Discussion by FerGT50** [earlephilhower/arduino-pico#2171] [Q&A] [answered]
Hello,
the default FIFO size for UART ports is 32 bytes.  I seem to remember from the Raspberry forum that this is the actual size of an internal hardware buffer.

But then we have setFIFOSize(), which apparently accepts whatever size_t value.

How does this work? Does it add a software buffer to the hardware buffer?
Is there any bad side effect to set it to a large value (say 1024 bytes)?

I have a use case where I have to receive data from a serial port set at 1'500'000 baud and don't want to lose data.

Many thanks

**Comment by earlephilhower** [comment]
Yes, this is a SW FIFO external to the HW one, driven by interrupts normally off of the HW-FIFO-not-empty signal.

**Comment by FerGT50** [comment]
Thank you. Do you have a suggestion about a sensible value for 1.5 Mbaud?

**Reply by earlephilhower** [reply, marked as answer]
Err, depends on the length of your message and how often you can service the serial port.  At some point you need to make the servicing of messages == the rate they're coming in or it'll overflow no matter how big the buffer.  Making it bigger shouldn't cause any issue, so if you have a 1K message @ 1.5MBaud every second, I'd just pop in a 1K buffer and be done with it...

Back of the envelope, 1.5MBaud = 150Kbyte/sec which at 133MHZ = ~900 CPU insns/byte (times two since there's a second core) so it doesn't seem too bad even on smaller buffers if you're only servicing the serial port.

**Reply by FerGT50** [reply]
It is a serial logger of sort: I am reading incoming data from the serial port, showing printable chars on a SPI display (with scrolling etc.) and writing to SD.
But I can always put the second core to use, yes, so I'll try 1 KB as a starting point.
Thank you!
